google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/DirectedReadTest.java

```diff
@@ -0,0 +1,251 @@
+      try (Connection connection = createConnection()) {
+        connection.setAutocommit(true);
+        connection.execute(
+            Statement.of(
+                String.format(
+                    "set %sdirected_read='%s'",
+                    getVariablePrefix(), DirectedReadOptionsUtil.toString(expected))));
+        executeQuery(connection);
+        assertDirectedReadOptions(expected);
+
+        mockSpanner.clearRequests();
+
+        // Reset to default.
+        connection.execute(
+            Statement.of(String.format("set %sdirected_read=''", getVariablePrefix())));
+        executeQuery(connection);
+        assertDirectedReadOptions(DirectedReadOptions.getDefaultInstance());
+
+        mockSpanner.clearRequests();
+      }
+    }
+  }
+
+  @Test
+  public void testDirectedReadIsIgnoredInReadWriteTransaction() {
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(false);
+      connection.setReadOnly(false);
+
+      connection.execute(
+          Statement.of(
+              String.format(
+                  "set %sdirected_read='%s'",
+                  getVariablePrefix(),
+                  DirectedReadOptionsUtil.toString(
+                      DirectedReadOptions.newBuilder()
+                          .setIncludeReplicas(
+                              IncludeReplicas.newBuilder()
+                                  .addReplicaSelections(
+                                      ReplicaSelection.newBuilder()
+                                          .setType(ReplicaSelection.Type.READ_WRITE)
+                                          .setLocation("us-west1")
+                                          .build())
+                                  .build())
+                          .build()))));
+      // This uses a read/write transaction, which will ignore any DirectedReadOptions.
+      executeQuery(connection);
+      assertDirectedReadOptions(DirectedReadOptions.getDefaultInstance());
+    }
+  }
+
+  @Test
+  public void testDirectedReadIsUsedInReadOnlyTransaction() {
+    DirectedReadOptions expected =
+        DirectedReadOptions.newBuilder()
+            .setIncludeReplicas(
+                IncludeReplicas.newBuilder()
+                    .addReplicaSelections(
+                        ReplicaSelection.newBuilder()
+                            .setType(ReplicaSelection.Type.READ_WRITE)
+                            .setLocation("us-west1")
+                            .build())
+                    .build())
+            .build();
+
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(false);
+      connection.setReadOnly(true);
+
+      connection.execute(
+          Statement.of(
+              String.format(
+                  "set %sdirected_read='%s'",
+                  getVariablePrefix(), DirectedReadOptionsUtil.toString(expected))));
+      // This uses a read/write transaction, which will ignore any DirectedReadOptions.
```

Comment:
Why would this use a RW transaction as connection. is readOnly mode?

Reply:
You're right, that's a copy-paste error. Fixed.